[CSS Fonts] The `from-font` value for CSS `font-size-adjust` is not computed correctly

shivamidow opened the pull request:
#### 89bdb0336c3b32a4130822ed976d858538e0d596
<pre>
[CSS Fonts] The `from-font` value for CSS `font-size-adjust` is not computed correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=259551">https://bugs.webkit.org/show_bug.cgi?id=259551</a>

Reviewed by Tim Nguyen.

When resolving styles, we used the parent-style font to compute the aspect ratio
for font-size-adjust: from-font because the first available font could not be
determined at that moment. Unfortunately, this causes the failure of
font-size-adjust-014.html. To make the test happy, we should defer resolving
the aspect value after determining a primary font. Also, FontDescriptionKey
should consider if the font-size-adjust has from-font, not simply checking if
it has a valid value when looking up a proper FontCascadeFonts.

For minimizing performance regression, this patch is updated as follows.
1. The resolving place changed from FontCascadeFonts::glyphDataForCharacter() to
   FontCascadeFonts::primaryFont(). Font-size-adjust:from-font is resolved when
   the primary font is created the first time.
2. Copying FontCascadeDescription is avoided. Instead, the source font description
   is directly modified.

* LayoutTests/TestExpectations:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKey::FontDescriptionKey):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::primaryFont):
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::operator bool const):
(WebCore::FontSizeAdjust::resolve const):
(WebCore::add):
(WebCore::operator&lt;&lt;):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::aspectValueOfPrimaryFont): Deleted.
* Source/WebCore/style/StyleFontSizeFunctions.h:

Canonical link: <a href="https://commits.webkit.org/269041@main">https://commits.webkit.org/269041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/583b4bd84cf99c4379db283d7b0cea448bc9df28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21414 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21722 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22467 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23276 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19846 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21655 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/25023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21969 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21040 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18554 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24128 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19403 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25724 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19540 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19612 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23572 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17110 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19415 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5118 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23655 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20003 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/platform/graphics/FontCascadeFonts.h:
```diff
@@ -148,6 +148,12 @@ inline const Font& FontCascadeFonts::primaryFont(const FontCascadeDescription& d
         }
+
+        auto fontSizeAdjust = description.fontSizeAdjust();
+        if (m_cachedPrimaryFont && fontSizeAdjust.isFromFont) {
```
Shouldn't `m_cachedPrimaryFont` be an `ASSERT` at this point?

shivamidow replied:
You are right. Is removing the m_cachedPrimaryFont check from the if condition OK? Or else, is putting an `ASSERT` expected?
```
ASSERT(m_cachedPrimaryFont);
auto fontSizeAdjust = description.fontSizeAdjust();
if (fontSizeAdjust.isFromFont) {
...
}
```

nt1m replied:
Yes, this is what I was thinking of.

shivamidow replied:
Thanks for the confirmation.

nt1m commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -319,7 +319,7 @@ class FontCascade : public CanMakeWeakPtr<FontCascade> {
     }
 
-    FontCascadeDescription m_fontDescription;
+    mutable FontCascadeDescription m_fontDescription;
```
Unfortunate we need to make the `FontCascadeDescription` just for `font-size-adjust: from-font`, but maybe @anttijk or @litherum has better ideas.

shivamidow replied:
Would you say more? Do you mean we need to create and use a clone of `FontCascadeDescription` for `font-size-adjust: from-font` in `FontCascade`? Even in that case, we may need `mutable` to avoid the `const` method chain.

nt1m replied:
I'll defer to Myles/Antti, also ok with addressing this in a follow up.

shivamidow replied:
O.K. Let's go with this change for now. I will follow up on other's comments if any.

nt1m reviewed (approved):
Looks good overall!

nt1m commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -45,6 +46,7 @@ struct FloatMarkableTraits {
 
 struct FontSizeAdjust {
     friend bool operator==(const FontSizeAdjust&, const FontSizeAdjust&) = default;
+    explicit operator bool() const { return value || isFromFont; }
```
I would prefer an isNone method than a custom book operator.

shivamidow replied:
My intention here is to check if `font-size-adjust` has either a number value or `from-font` with a single method. Otherwise, we need to check two times if FontSizeAdjust has a value, something like `fontSizeAdjust.value || fontSizeAdjust.isFromFont` above. (e.g., FontCascadeCache.h:123,  https://github.com/WebKit/WebKit/pull/18425/files#diff-3dd26ef157ca99d4dd427524bfd9a08c87d6fe42275ad1ed9fde359296067cffL123)

Also, isNone() sounds a bit confusing to me. It could be understood as `font-size-adjust: none`.

Anyhow, I have a plan to clean up the value type here. We can discuss it more in a follow-up patch if you don't mind.

nt1m replied:
I was thinking isNone to represent "is the value none", and then you could check the negation of that to see if from-font or a value was specified.

nt1m replied:
So everywhere you currently check for `fontSizeAdjust`, you would check for `!fontSizeAdjust.isNone()` instead. I think that's nicer and more explicit.

shivamidow replied:
I see. But isNone() sounds like a confusing name. What about isEmpty() or hasValue()?

nt1m replied:
Why is it a confusing name? Isn't there a corresponding `font-size-adjust: none` value? My intention is that it maps to that. I'm not a fan of `hasValue()` because technically `from-font` has no value when unresolved. I prefer `isNone` over `isEmpty` for the parallel with the CSS value.

shivamidow replied:
Because `font-size-adjust` can have `none` as its value. https://www.w3.org/TR/css-fonts-5/#font-size-adjust-prop

nt1m replied:
Yes, that is my point. isNone would represent when font-size-adjust is set to none

nt1m replied:
@shivamidow To be clear, I think `isNone()` should be defined like: `!fontSizeAdjust.value && !fontSizeAdjust.isFromFont` and then everywhere you currently check for `fontSizeAdjust`, you change it to `!fontSizeAdjust.isNone()`

shivamidow replied:
O.K. I will follow up. Stay tuned. Thanks.

shivamidow committed:
[CSS Fonts] The `from-font` value for CSS `font-size-adjust` is not computed correctly
https://bugs.webkit.org/show_bug.cgi?id=259551

Reviewed by Tim Nguyen.

When resolving styles, we used the parent-style font to compute the aspect ratio
for font-size-adjust: from-font because the first available font could not be
determined at that moment. Unfortunately, this causes the failure of
font-size-adjust-014.html. To make the test happy, we should defer resolving
the aspect value after determining a primary font. Also, FontDescriptionKey
should consider if the font-size-adjust has from-font, not simply checking if
it has a valid value when looking up a proper FontCascadeFonts.

For minimizing performance regression, this patch is updated as follows.
1. The resolving place changed from FontCascadeFonts::glyphDataForCharacter() to
   FontCascadeFonts::primaryFont(). Font-size-adjust:from-font is resolved when
   the primary font is created the first time.
2. Copying FontCascadeDescription is avoided. Instead, the source font description
   is directly modified.

* LayoutTests/TestExpectations:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKey::FontDescriptionKey):
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::primaryFont):
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::operator bool const):
(WebCore::FontSizeAdjust::resolve const):
(WebCore::add):
(WebCore::operator<<):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontSizeAdjust):
* Source/WebCore/style/StyleFontSizeFunctions.cpp:
(WebCore::Style::aspectValueOfPrimaryFont): Deleted.
* Source/WebCore/style/StyleFontSizeFunctions.h:

Canonical link: https://commits.webkit.org/269041@main
